weed/filer/stream.go

```diff
@@ -33,16 +34,32 @@ func StreamContent(masterClient wdclient.HasLookupFileIdFunction, w io.Writer, c
 		fileId2Url[chunkView.FileId] = urlStrings
 	}
 
-	for _, chunkView := range chunkViews {
+	for idx, chunkView := range chunkViews {
 
 		urlStrings := fileId2Url[chunkView.FileId]
 
-		data, err := retriedFetchChunkData(urlStrings, chunkView.CipherKey, chunkView.IsGzipped, chunkView.IsFullChunk(), chunkView.Offset, int(chunkView.Size))
+		// Pre-check all chunkViews urls
+		gErr := new(errgroup.Group)
+		if len(chunkViews) > 1 && idx == 0 {
+			CheckAllChunkViews(chunkViews[1:], &fileId2Url, gErr)
```

Comment:
seems a lot of overhead here.

Reply:
> seems a lot of overhead here.

here is log, check requests with bytes=0-1 processed in parallel with the receipt of the first chunk
```
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /3,16bed1797b 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /3,1994fd057f bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /5,1d032c3d04 bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /4,176f4d5358 bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /7,1c3133458c bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /4,1ba19850c6 bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /5,1f72f56cc2 bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /4,21b3aabf73 bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /5,1e6d572d2b bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /3,22136b5ac6 bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /6,258329f089 bytes=0-1
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /3,18e9cbf3f2
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /7,1a9119ffd1 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /5,1d032c3d04 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /5,1f72f56cc2 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /4,20a3a041ac 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /3,23f7514b68 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /4,24cade9bbf 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /4,176f4d5358 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /3,1994fd057f 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /4,1ba19850c6 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /7,1c3133458c 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /5,1e6d572d2b 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /4,21b3aabf73 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /3,22136b5ac6 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /6,258329f089 
volume_1  | I0315 18:29:33     1 volume_server_handlers_read.go:30] GET /3,1595545117 
```

Reply:
so one request becomes two requests?

Reply:
yes it may slightly increase the utilization of the CPU. But on the other hand, the http connection warms up

Reply:
It is "free" when the system load is low, but it is costly when the load is high. 

It is a coding problem and should not rely on extra network round trip.

Reply:
Let's move behind the option then? 
In fact, s3 makes range requests, if they match the filler chunk size, then there will always be one request.
So it's more like a reassurance.

Reply:
This does not worth an option. I think this is ok to return a 200 status code, promising a good response, but failed in the middle.

Reply:
> This does not worth an option. I think this is ok to return a 200 status code, promising a good response, but failed in the middle.

It is necessary to somehow distinguish normal responses from broken ones in the middle in monitoring. Here you can add a metric and follow it.
If there are many such requests, it will be possible to take action.

We are also looking forward to the opportunity to repeat the request for 5xx on the upstream

Reply:
There is also an option to move the check only for HEAD requests ?